### ISSUE TITLE
Problem with low metrics

### ISSUE THREAD

**Issue by panagiotamoraiti** [MultimediaTechLab/YOLO#197]
Hello,

I've been working on running YOLOv9 on my custom dataset with 10 classes. After repeatedly training with poor results, metrics consistently below 1%, I discovered that reducing the learning rate to 0.0001 and disabling most data augmentations significantly improved performance. The only augmentation I kept was a horizontal flip with a probability of 0.2.

Since making these changes, my metrics have started to improve. After just 10 epochs, I'm now seeing around 50% AP and AR. If you're struggling with low metrics, I highly recommend trying these adjustments.

**Comment by Rhroan** [comment]
My experence ， when few class , use Mosaic and RandomCrop is deficult to train  especially when there is only one category.

I suggest adding 
RandomBrightness,
 RandomContrast,
 RandomSaturation to image enhancement training

![Image](https://github.com/user-attachments/assets/911a624b-9ee1-448c-8801-ff682afaf136)

![Image](https://github.com/user-attachments/assets/34dd25b4-fdca-432e-85e8-015e5e66a901)

**Comment by mbegue-cr** [comment]
THANK YOU!  This Issue and its comments saved a lot of headache.

**Comment by Noah0376** [comment]
How do you add Brightness, contrast, and saturation augmentation? I don't have those as an option for the training yaml

Edit: Also, if there are any other data augmentation or training settings you are willing to share that you used to get these kind of results, I'd appreciate the help.

**Comment by BitCode-in** [comment]
Hi everyone,

I had the exact same issue with low metrics on a custom dataset (small objects on complex backgrounds) and wanted to share what actually fixed it for me.

@panagiotamoraiti Dropping the LR to 0.0001 and removing Mosaic/MixUp gives a quick spike in AP, but it's usually an overfitting trap. Without Mosaic, the model just memorizes your specific training images. Also, if you're using SGD, an LR of 0.0001 is way too low to properly update weights.

@Noah0376 If your repo doesn't support color augs from the yaml, you have to write a custom transform using torchvision. Here is a quick example for brightness that you can add to your data augmentation script:

```python
import torchvision.transforms.functional as TF
import numpy as np
import torch

class RandomBrightness:
    def __init__(self, prob=0.5, factor_range=[0.8, 1.2]):
        self.prob = prob
        self.factor_range = factor_range

    def __call__(self, image, boxes):
        if torch.rand(1) < self.prob:
            factor = np.random.uniform(self.factor_range[0], self.factor_range[1])
            image = TF.adjust_brightness(image, factor)
        return image, boxes
```
Just make sure your yaml parser unpacks the dict parameters correctly so `prob` doesn't accidentally get passed as a config object.

To save you all some time, here is my complete train.yaml. Switching from SGD to AdamW (with lr 0.001) fixed the convergence speed for me. 

But I also want to highly recommend looking at the scheduler block. Switching to CosineAnnealingLR instead of Linear was another massive leap for my metrics.

```yaml
task: train

defaults:
  - validation: ../validation

epoch: 500

data:
  batch_size: 16
  equivalent_batch_size: 64
  image_size: ${image_size} 
  cpu_num: ${cpu_num}
  shuffle: True
  pin_memory: True
  
  data_augment:
    Mosaic: 1.0
    MixUp: 0.1
    HorizontalFlip: 0.5
    RandomCrop: 0.0
    VerticalFlip: 0.0
    
    RandomBrightness:
      prob: 0.5
      factor_range: [0.6, 1.4]
    RandomContrast:
      prob: 0.5
      factor_range: [0.6, 1.4]
    RandomSaturation:
      prob: 0.5
      factor_range: [0.6, 1.4]
    
    RemoveOutliers: 1e-8

optimizer:
  type: AdamW
  args:
    lr: 0.001
    weight_decay: 0.05
    betas: [0.9, 0.999]

loss:
  objective:
    BCELoss: 0.5
    BoxLoss: 7.5
    DFLoss: 1.5
  aux: 0.25
  matcher:
    iou: CIoU
    topk: 10
    factor:
      iou: 6.0
      cls: 0.5

scheduler:
  type: CosineAnnealingLR
  warmup:
    epochs: 3.0
  args:
    T_max: ${task.epoch}
    eta_min: 0.0001

ema:
  enable: true
  decay: 0.9999
```

Hope this helps you guys get those metrics up.
